File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys
from collections import defaultdict
sys.setrecursionlimit(100000)

input = sys.stdin.readline
n,m = map(int,input().split())
parent = defaultdict(int)

for i in range(1,n+1):
    parent[i] = i

def find(target):
    if target == parent[target]:
        return target
    
    parent[target] = find(parent[target])
    return parent[target]

def union(a,b):
    a = find(a)
    b = find(b)

    if a < b:
        parent[b] = a
    else:
        parent[a] = b

for _ in range(m):
    op,a,b = map(int,input().split())
    if op == 0:
        union(a,b)
    else:
        if find(a) == find(b):
            print("YES")
        else:
            print("NO")

File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <iostream>
using namespace std;
int n,m;
int parent[1000001];
int find(int a){
    if(parent[a] == a) return a;
    parent[a] = find(parent[a]);
    return parent[a];
}
bool merge(int a, int b){
    a = find(a);
    b = find(b);
    if(a==b) return false;
    if(a<b) parent[b] = a;
    else parent[a] = b;
    return true;
}
int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin >> n >> m;
    for(int i=1; i<=n; i++) parent[i] = i;
    for(int i=0; i<m; i++){
        int op,a,b;
        cin >> op >> a >> b;
        if(op==0){
            merge(a,b);
        }
        else{ // op==1
            if(find(a) != find(b)) cout <<"NO\n";
            else cout << "YES\n";
        }
    }
}